File: src/text_cleaner.py
"""
text_cleaner.py
───────────────
Display-time text normalization for article names.
Raw data from Excel is NEVER modified — corrections are applied only when
rendering to the preview canvas or generating PDF output.

Two passes:
  1. Whitespace normalization — collapse multiple spaces, trim, and insert a
     space at letter→digit boundaries (e.g. "bouteille1x" → "bouteille 1x").
  2. Brand spell correction — regex word-boundary replacement for known
     misspellings of popular Belgian drink brands.
     Word boundaries prevent "red bul" from partially matching "red bull".
"""

import re

# ── Brand corrections ────────────────────────────────────────────────────────
# Key   : lowercase misspelling / common variant
# Value : canonical brand spelling
_CORRECTIONS = {
    # Red Bull
    "redbull":          "Red Bull",
    "red-bull":         "Red Bull",
    "red bul":          "Red Bull",
    "redbul":           "Red Bull",
    # Coca-Cola
    "cocacola":         "Coca-Cola",
    "coca cola":        "Coca-Cola",
    "coka cola":        "Coca-Cola",
    "coka-cola":        "Coca-Cola",
    # Pepsi
    "pepsy":            "Pepsi",
    # Fanta
    "phanta":           "Fanta",
    # Sprite
    "sprit":            "Sprite",
    # Monster Energy
    "monster energy":   "Monster Energy",
    "monsteur":         "Monster",
    # Heineken
    "heiniken":         "Heineken",
    "heinekin":         "Heineken",
    "heinekn":          "Heineken",
    "heinneken":        "Heineken",
    # Jupiler
    "jupiller":         "Jupiler",
    "jupilier":         "Jupiler",
    # Hoegaarden
    "hoegarden":        "Hoegaarden",
    # Stella Artois
    "stela artois":     "Stella Artois",
    "stella artoi":     "Stella Artois",
    # Duvel
    "duvell":           "Duvel",
    # Leffe
    "lefe":             "Leffe",
    "leff":             "Leffe",
    # Chimay
    "chimaye":          "Chimay",
    # Desperados
    "desperado":        "Desperados",
    # Kronenbourg
    "kronenburg":       "Kronenbourg",
    "kronenbug":        "Kronenbourg",
    # Corona
    "korona":           "Corona",
    # Budweiser
    "budweizer":        "Budweiser",
    "budwieser":        "Budweiser",
    # Schweppes
    "schwepps":         "Schweppes",
    "schweps":          "Schweppes",
    "shweppes":         "Schweppes",
    # Perrier
    "perier":           "Perrier",
    # Evian
    "evien":            "Evian",
    # San Pellegrino
    "san pellegrino":   "San Pellegrino",
    "san-pellegrino":   "San Pellegrino",
    # Lipton Ice Tea
    "lipton ice tea":   "Lipton Ice Tea",
    # Tropicana
    "tropicanna":       "Tropicana",
    # Minute Maid
    "minut maid":       "Minute Maid",
}

# Pre-compile regex patterns with word boundaries, longest pattern first
# (?<!\w)...\b ensures we don't partial-match "red bul" inside "red bull"
_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r'(?<!\w)' + re.escape(wrong) + r'(?!\w)', re.IGNORECASE), right)
    for wrong, right in sorted(_CORRECTIONS.items(), key=lambda kv: len(kv[0]), reverse=True)
]


def clean_article(text: str) -> str:
    """
    Normalize and spell-correct an article name for display/print.
    Source data is never modified — call this only at render time.
    """
    if not text:
        return text

    # Pass 1: whitespace normalization
    text = text.strip()
    text = re.sub(r'\s+', ' ', text)                  # collapse runs of spaces
    text = re.sub(r'([a-zA-Z])(\d)', r'\1 \2', text)  # letter→digit boundary

    # Pass 2: brand corrections
    for pattern, right in _PATTERNS:
        text = pattern.sub(right, text)

    return text

File: web/src/lib/textCleaner.js
/**
 * Text cleaner: normalises whitespace, inserts spaces at letter→digit
 * boundaries, and corrects common brand-name misspellings.
 *
 * Ported from Python text_cleaner.py
 */

const CORRECTIONS = {
  'redbull': 'Red Bull',
  'red-bull': 'Red Bull',
  'red bul': 'Red Bull',
  'redbul': 'Red Bull',
  'cocacola': 'Coca-Cola',
  'coca cola': 'Coca-Cola',
  'coka cola': 'Coca-Cola',
  'coka-cola': 'Coca-Cola',
  'pepsy': 'Pepsi',
  'phanta': 'Fanta',
  'sprit': 'Sprite',
  'monster energy': 'Monster Energy',
  'monsteur': 'Monster',
  'heiniken': 'Heineken',
  'heinekin': 'Heineken',
  'heinekn': 'Heineken',
  'heinneken': 'Heineken',
  'jupiller': 'Jupiler',
  'jupilier': 'Jupiler',
  'hoegarden': 'Hoegaarden',
  'stela artois': 'Stella Artois',
  'stella artoi': 'Stella Artois',
  'duvell': 'Duvel',
  'lefe': 'Leffe',
  'leff': 'Leffe',
  'chimaye': 'Chimay',
  'desperado': 'Desperados',
  'kronenburg': 'Kronenbourg',
  'kronenbug': 'Kronenbourg',
  'korona': 'Corona',
  'budweizer': 'Budweiser',
  'budwieser': 'Budweiser',
  'schwepps': 'Schweppes',
  'schweps': 'Schweppes',
  'shweppes': 'Schweppes',
  'perier': 'Perrier',
  'evien': 'Evian',
  'san pellegrino': 'San Pellegrino',
  'san-pellegrino': 'San Pellegrino',
  'lipton ice tea': 'Lipton Ice Tea',
  'tropicanna': 'Tropicana',
  'minut maid': 'Minute Maid',
};

// Build regex patterns sorted longest-first so multi-word keys match before
// shorter substrings.  Each pattern uses word boundaries and is case-insensitive.
const patterns = Object.keys(CORRECTIONS)
  .sort((a, b) => b.length - a.length)
  .map((key) => ({
    regex: new RegExp(`\\b${key.replace(/-/g, '\\-')}\\b`, 'gi'),
    replacement: CORRECTIONS[key],
  }));

/**
 * Clean an article description:
 *  1. Trim, collapse whitespace, insert space at letter→digit boundary
 *  2. Apply brand-name corrections (case-insensitive)
 *
 * @param {string|null|undefined} text
 * @returns {string}
 */
export function cleanArticle(text) {
  if (!text) return '';

  // Pass 1: trim + collapse whitespace + letter→digit boundary
  let result = text.trim().replace(/\s+/g, ' ').replace(/([a-zA-Z])(\d)/g, '$1 $2');

  // Pass 2: brand corrections
  for (const { regex, replacement } of patterns) {
    result = result.replace(regex, replacement);
  }

  return result;
}
